### ISSUE TITLE
Automapper sets up wrong mappings

### ISSUE THREAD

**Issue by PovilasVytautasC** [AutoMapper/AutoMapper#431]
To Help you understand the problem better, I will explain the relations between objects before presenting the problem.

There is a collection of reservations. Each reservation is related to a single ExamDescription object. Each exam description has a collection of ExamDescriptionValue objects and is related to a single ExamType object. Each ExamType object has a collection of ExamTypeValue objects.

Each "value" object represents the object in a different language. for exam ExamDescriptionValue object contains information in a specific language about the ExamDescription object.

That is it for the DB objects. Now for business logic classes (the viewmodels)

Each ViewModel class contain all properties of the corresponding DB object. Also, if the view model represents a multilingual object, the direct mapping goes ot the value object. For exam There is 1 ExamType and 3 ExamTypeValue DB objects. This would be mapped into a 3 ExamTypeViewModel object, each containing information of every ExamTypeValue object, plus all information from the ExamType object.

So now that the data architecture is a bit more clear - onward to the problem itself :)

So I am pulling out my reservations from the DB and mapping it to the corresponding ReservationViewModel class like this:

```
        Mapper.CreateMap<Models.DbClasses.Reservation, ViewModels.Reservations.ReservationViewModel>();
```

Pretty straight-forward. However, each ReservationviewModel has a reference to a collection of ExamDescriptionViewModels (exam descriptions in all languages). To have all data in the view model, after the mapping is done (using queryable projections) I ahve written some code myself to loop through a collection of mapped ReservationViewModels and for each reservation manually add examDescriptions. This is the method that achieves this:

```
public List<ReservationViewModel> AddExamDescritions(IQueryable<ReservationViewModel> Reservations)
        {
            List<ReservationViewModel> ReservationList = Reservations.ToList();
            List<string> ExamDescriptionIDs = ReservationList.Select(item => item.ExamDescriptionId).ToList();
            List<Models.DbClasses.ExamDescription> ExamDescriptionsDb = UnitOfWork.ExamDescriptionsRepository.Get(filter: item => ExamDescriptionIDs.Contains(item.ID)).ToList();

            foreach (ReservationViewModel Reservation in ReservationList)
            {
                Reservation.ExamDescriptions = Mapper.Map<Models.DbClasses.ExamDescription, List<ExamDescriptionViewModel>>(ExamDescriptionsDb.Single(item => item.ID == Reservation.ExamDescriptionId));
            }

            return ReservationList;
        }
```

The ExamDescription -> List<ExamDescriptionViewModel> mapping is done as follows:

```
Mapper.CreateMap<Models.DbClasses.ExamDescription, List<ViewModels.Exams.ExamDescriptionViewModel>>()
                .AfterMap((source, destination) => Mapper.Map<List<Models.DbClasses.ExamDescriptionValue>, List<ViewModels.Exams.ExamDescriptionViewModel>>(source.ExamDescriptionValues.ToList(), destination));
```

And this is where the problem happens: Automapper throws an exception "Mapping types:
ExamTypeValue_A793753CF51E1AC960B2D858C065FC11ACF92146A58201E7EAE6CF1BB1A3B430 -> Boolean
System.Data.Entity.DynamicProxies.ExamTypeValue_A793753CF51E1AC960B2D858C065FC11ACF92146A58201E7EAE6CF1BB1A3B430 -> System.Boolean

Destination path:
List`1[0].ExamType.ExamType.Used.Used

Source value:
System.Data.Entity.DynamicProxies.ExamTypeValue_A793753CF51E1AC960B2D858C065FC11ACF92146A58201E7EAE6CF1BB1A3B430"

So this tells me that I am trying to map ExamTypeValue object to a property of the ExamTypeViewModel object, however I am not doing this. The mapping for "ExamTypeValue" -> "ExamTypeViewModel" is achieve via the following code:

```
            Mapper.CreateMap<Models.DbClasses.ExamTypeValue, ViewModels.Classificators.ExamTypeViewModel>()
                .ForMember(item => item.ID, property => property.MapFrom(item => item.ExamTypeId))
                .ForMember(item => item.ExamTypeValueId, property => property.MapFrom(item => item.ID))
                .ForMember(item => item.Code, property => property.MapFrom(item => item.ExamType.Code))
                .ForMember(item => item.FileId, property => property.MapFrom(item => item.ExamType.FileId))
                .ForMember(item => item.File, property => property.MapFrom(item => item.ExamType.File))
                .ForMember(item => item.Used, property => property.MapFrom(item => item.ExamType.ExamDescriptions
                    .Any(examDescription => examDescription.Reservations
                    .Any(reservation => reservation.Registrations
                    .Any(registration => registration.Exams.Any())))));
```

Basically, the "Used" property shows, that the object is used in other parts of the system, and so it just checks is some grand-grandchildren exists. I would continue digging in the code myself if not for one detail: in other parts of the system, the mapping actually works, automapper correctly assigns the "used" value and does not give me any exceptions.

Also, I've checked the DB objects that I am passing as source values. They are correct. All relations are there, all objects are fetched correctly. No information is missing.

So could anyone, please give me an idea what is causing this bizzare behaviour? If needed I can provide with more details. Than you in advance for reading all... this.
